src/frequenz/channels/select.py

```diff
@@ -20,15 +20,49 @@
 T = TypeVar("T")
 
 
-@dataclass
+@dataclass(frozen=True)
 class _Selected:
     """A wrapper class for holding values in `Select`.
 
     Using this wrapper class allows `Select` to inform user code when a
-    receiver gets closed.
+    receiver gets closed or raises an exception.
     """
 
-    inner: Optional[Any]
+    _result: Any
+
+    @property
+    def result(self) -> Any:
+        """The result from a selected receiver or None if the channel was closed.
+
+        Returns: the value of the receiver or None if the channel was closed.
+
+        Raises:
+            Exception: if the value is an exception.
+        """
+        if isinstance(self._result, Exception):
+            raise self._result
+        return self._result
```

Comment:
> I guess this should be value instead of result? Result would suggest some operation was performed on some other data to produce the result.

Yeah, I wasn't sure, I like that this follows the `Future` interface, but it is not really a future (so the interface is much more limited than a `Future`). We could name it `value`. I was also thinking about `message`. It seems we are not using very consistent naming, sometimes it says *value*, sometimes *message*.

Maybe we can say a *message* can carry a *value* or an *exception*?

Reply:
> Maybe we can say a message can carry a value or an exception?

yes,  that makes sense.

Reply:
> Also, why not just have _result and _exception as instance variables? Else, when users call exception first, we'd do an if isinstance, and once again when they call result. Not a big cost, but using variables would also mean simpler code as well.

I could do that too, I have no strong reason, I was just mimicking what comes in the channel. I guess the most typical use case will be using `result` and trying to catch the exception and `result` will need to check if there was no exception anyway.

Reply:
oh I see,  `result` raises.  ok,  then its probably better like this.